piptools/build.py

```diff
@@ -34,23 +41,93 @@ def get_all(self, name: str, failobj: _T) -> list[Any] | _T:
             ...
 
 
+@dataclass
+class StaticProjectMetadata:
+    extras: tuple[str, ...]
+    requirements: tuple[InstallRequirement, ...]
+
+
 @dataclass
 class ProjectMetadata:
     extras: tuple[str, ...]
     requirements: tuple[InstallRequirement, ...]
     build_requirements: tuple[InstallRequirement, ...]
 
 
+def maybe_statically_parse_project_metadata(
+    src_file: pathlib.Path,
+) -> StaticProjectMetadata | None:
+    """
+    Return the metadata for a project, if it can be statically parsed from pyproject.toml.
+
+    This function is typically significantly faster than invoking a build backend.
+    Returns None if the project metadata cannot be statically parsed.
+    """
+    if src_file.name != PYPROJECT_TOML:
+        return None
+
+    try:
+        with open(src_file, "rb") as f:
+            pyproject_contents = tomllib.load(f)
+    except tomllib.TOMLDecodeError:
+        return None
+
+    # Not valid PEP 621 metadata
+    if (
+        "project" not in pyproject_contents
+        or "name" not in pyproject_contents["project"]
+    ):
```

Comment:
I wonder if we could use `validate-pyproject` here?

Reply:
Possibly? What kind of situation are you thinking the extra dependency helps with?

Reply:
According to the [schemas page](https://validate-pyproject.readthedocs.io/en/latest/schemas.html), this library validates the `pyproject.toml`  structure according to PEP 621 quite thoroughly.

This was also mostly a note to myself. Utilizing this library would enable failing fast on invalid `pyproject.toml` files? There might be some performance issues, though.